TrackingTools/TrackFitters/plugins/KFFittingSmoother.h

```diff
@@ -0,0 +1,435 @@
+								     *theSmoother,
+								     *this)
+					       );
+    }
+    
+    
+    virtual void setHitCloner(TkCloner const * hc) override {
+      theFitter->setHitCloner(hc);
+      theSmoother->setHitCloner(hc);
+    }
+    
+    KFFittingSmoother(const TrajectoryFitter& aFitter,
+		      const TrajectorySmoother& aSmoother,
+		      KFFittingSmootherParam const & other) :
+      KFFittingSmootherParam(other),
+      theFitter(aFitter.clone()),
+      theSmoother(aSmoother.clone())
+	{}
+    
+    
+    
+    Trajectory smoothingStep(Trajectory && fitted) const {
+      if (theEstimateCut>0) {
+	// remove "outlier" at the end of Traj
+	while (!fitted.empty() &&
+	       ( !fitted.lastMeasurement().recHitR().isValid()
+		 || ( fitted.lastMeasurement().recHitR().det()!=nullptr && fitted.lastMeasurement().estimate()>theEstimateCut)
+		 )
+	       )
+	  fitted.pop();
+	if (fitted.foundHits() < theMinNumberOfHits ) return Trajectory();
+      }
+      return theSmoother->trajectory(fitted);
+    }
+    
+  private:
+    
+    const std::unique_ptr<TrajectoryFitter> theFitter;
+    const std::unique_ptr<TrajectorySmoother> theSmoother;
+    
+    
+    /// Method to check that the trajectory has no NaN in the states and chi2
+    static bool checkForNans(const Trajectory &theTraj);
+
+    friend class  KFFittingSmootherESProducer;
+    
+  };
+
+
+  // #define VI_DEBUG
+  
+#ifdef VI_DEBUG
+#define DPRINT(x) std::cout << x << ": "
+#define PRINT std::cout
+#else
+#define DPRINT(x) LogTrace(x)
+#define PRINT LogTrace("")
+#endif
+
+
+using namespace std;
+
+Trajectory KFFittingSmoother::fitOne(const Trajectory& t, fitType type) const {
+  if (!t.isValid() ) return Trajectory();
+  return smoothingStep(theFitter->fitOne(t,type));
+}
+
+bool KFFittingSmoother::checkForNans(const Trajectory & theTraj)
+{
+  if (edm::isNotFinite(theTraj.chiSquared ())) return false;
+  auto const & vtm = theTraj.measurements();
+  for (auto const &  tm : vtm) {
+    if (edm::isNotFinite(tm.estimate())) return false;
+    auto const & v = tm.updatedState ().localParameters ().vector();
+    for (int i=0;i<5;++i) if (edm::isNotFinite(v[i])) return false;
+    auto const & m = tm.updatedState ().curvilinearError ().matrix();
+    for (int i=0;i<5;++i)
+      for (int j=0;j<(i+1);++j) if (edm::isNotFinite(m(i,j))) return false;
+  }
+  return true;
+}
+
+Trajectory KFFittingSmoother::fitOne(const TrajectorySeed& aSeed,
+				  const RecHitContainer& hits,
+				  const TrajectoryStateOnSurface& firstPredTsos,
+				  fitType type) const
+{
+  LogDebug("TrackFitters") << "In KFFittingSmoother::fit";
+
+  if ( hits.empty() ) return Trajectory();
+
+  bool hasoutliers=false;
+
+  RecHitContainer myHits = hits;
+  Trajectory tmp_first;
+  // bool firstTry=true;
+
+  //call the fitter
+  Trajectory smoothed  = smoothingStep(theFitter->fitOne(aSeed, myHits, firstPredTsos));
+
+  
+  do {
+
+    //if no outliers the fit is done only once
```

Comment:
please cleanup the commented out code.
If it's needed, better hide by some *DEBUG ifdef